# Day 2 Reading Journal

This journal includes several required exercises, but it is meant to encourage active reading more generally.  You should use the journal to take detailed notes, catalog questions, and explore the content from Think Python deeply.

Reading: Think Python Chapter 6.1-6.4, 7

**Due: Tuesday, January 24 at 12 noon**



## [Chapter 6.1-6.4](http://greenteapress.com/thinkpython2/html/thinkpython2007.html)

Note: the exercise numbers below match up with the reading for ease of cross referencing.

**Quick check:** What is a _fruitful function_?

A function that returns a value

### Chapter 6.1 
Write a `compare` function that returns `1` if `x > y`, `0` if `x == y`, and `-1` if `x < y`.

In [ ]:
def compare_funk():
    x = int(input("x"))
    y = int(input("y"))
    
    if x > y:
        return(1)
    elif x == y:
        return(0)
    elif x < y:
        return(-1)
    
compare_funk()
        
        

**Quick check:** Modify the `absolute_value` function so that it returns the correct result for all integer inputs.

In [ ]:
def absolute_value():
    x = int(input("input an integer"))
    
    if x < 0:
        return -x
    if x > 0:
        return x
    
absolute_value()

### Chapter 6.2 
Use incremental development to write a function called `hypotenuse` that returns the length of the hypotenuse of a right triangle given the lengths of the two legs as arguments. Record each stage of the development process as you go, using as many cells as you need.

In [ ]:
def hypotenuse():
    
    leg1 = int(input("length of leg 1"))
    leg2 = int(input("length of leg 2"))
    
    return(leg1, leg2)
        
hypotenuse()

In [ ]:
def hypotenuse():
    
    leg1 = int(input("length of leg 1"))
    leg2 = int(input("length of leg 2"))
    
    import math
    
    hypo = math.sqrt(leg1**2 + leg2**2)
    
    return(hypo)

hypotenuse()

# NOTE
### Incremental Development

The process of developing a large code by slowly adding processes.

### Chapter 6.4
Write a function `is_between(x, y, z)` that returns `True` if `x ≤ y ≤ z` or `False` otherwise.

In [ ]:
def is_between(x,y,z):

    if y >= x:
        check=1
    if y <= z:
        check = check+1
    
    if check == 2:
        return "True"
    else:
        return "False"

is_between(124,222,3313)

# NOTE
### Assigning Variables

Variables can be reassigned and depending on how you format it you can play around with assigning numbers and other variables in combination.

## [Chapter 7](http://greenteapress.com/thinkpython2/html/thinkpython2008.html)



**Quick check:** How do you test for equality in Python?

In [ ]:
Double equal sign, ==

**Challenge:** Prove/disprove the Collatz conjecture :)

lol

### Chapter 7.5
Encapsulate the loop from Section 7.5 in a function called `square_root` that takes `a` as a parameter, chooses a reasonable value of `x`, and returns an estimate of the square root of `a`.

In [3]:
def square_root(a):

    x = a/2
    y = (x + a/x) / 2
    
    x=y
    y = (x + a/x) / 2
    
    x=y
    y = (x + a/x) / 2
    
    return y
   
square_root(9)

3.000015360039322

### Exercise 7.1  
To test the square root algorithm you developed in Exercise 2, you could compare it with Python's `math.sqrt` function. Write a function named `test_square_root` that prints a table like this:

```python
1.0 1.0           1.0           0.0
2.0 1.41421356237 1.41421356237 2.22044604925e-16
3.0 1.73205080757 1.73205080757 0.0
4.0 2.0           2.0           0.0
5.0 2.2360679775  2.2360679775  0.0
6.0 2.44948974278 2.44948974278 0.0
7.0 2.64575131106 2.64575131106 0.0
8.0 2.82842712475 2.82842712475 4.4408920985e-16
9.0 3.0           3.0           0.0
```

The first column is a number, `a`; the second column is the square root of a computed with the function from Section 7.5; the third column is the square root computed by `math.sqrt`; the fourth column is the absolute value of the difference between the two estimates.

# NOTE

The function works until it gets to 5 and then it runs a bunch of errors and I'm not quite sure why.

In [13]:
def test_square_root(n):
    
        
    x = n/2
    y = (x + n/x) / 2
    
    x=y
    y = (x + n/x) / 2
    
    x=y
    y = (x + n/x) / 2
    
    import math
    
    real = math.sqrt(n)
    
    subtract = (y-real)
    
    if subtract > 0:
        absolute = subtract
    elif subtract<0:
        absolute = -subtract
    
    print(n,y,real, absolute)

    n = n-1
    
    if n>0:
        test_square_root(n)
    else:
        pass
   
test_square_root(9)

9 3.000015360039322 3.0 1.5360039321787866e-05
8 2.8284313725490193 2.8284271247461903 4.24780282903825e-06
7 2.6457520483808037 2.6457513110645907 7.37316212973127e-07
6 2.4494897959183675 2.449489742783178 5.313518958161012e-08
5 2.2360679779158037 2.23606797749979 4.1601389000334166e-10


UnboundLocalError: local variable 'absolute' referenced before assignment

### Challenge: Exercise 7.3  (optional)
The mathematician Srinivasa Ramanujan found an infinite series that can be used to generate a numerical approximation of
$\frac{1}{\pi}$:

$$\frac{1}{\pi} = \frac{2 \sqrt{2}}{9801} \sum_{k=0}^{\infty} \frac{(4k)! (1103+26390k)}{(k!)^4 396^{4k}}$$
 
Write a function called `estimate_pi` that uses this formula to compute and return an estimate of $\pi$. It should use a `while` loop to compute terms of the summation until the last term is smaller than `1e-15` (which is Python notation for $10^{−15}$). You can check the result by comparing it to `math.pi`.

[Possible solution](http://thinkpython.com/code/pi.py) (give the exercise a try before viewing the solution)

## Quick poll

About how long did you spend working on this Reading Journal?

1.5 hours

## Reading Journal feedback

Have any comments on this Reading Journal? Feel free to leave them below and we'll read them when you submit your journal entry. This could include suggestions to improve the exercises, topics you'd like to see covered in class next time, or other feedback.

If you have Python questions or run into problems while completing the reading, you should post them to Piazza instead so you can get a quick response before your journal is submitted.